cpmpy/solvers/cpo.py

```diff
@@ -0,0 +1,491 @@
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_template
+"""
+
+class CPM_cpo(SolverInterface):
+    """
+    Interface to CP Optimizers API
+
+    Requires that the 'docplex' python package is installed:
+    $ pip install docplex
+
+    docplex documentation:
+    https://ibmdecisionoptimization.github.io/docplex-doc/
+
+    You will also need to install CPLEX Optimization Studio from IBM's website.
+    There is a free community version available.
+    https://www.ibm.com/products/ilog-cplex-optimization-studio
+
+    See detailed installation instructions at:
+    https://www.ibm.com/docs/en/icos/22.1.2?topic=2212-installing-cplex-optimization-studio
+
+    Creates the following attributes (see parent constructor for more):
+    - cpo_model: object, CP Optimizers model object
+    """
+
+    @staticmethod
+    def supported():
+        return CPM_cpo.installed() and CPM_cpo.executable_installed()
+
+    @staticmethod
+    def installed():
+        # try to import the package
+        try:
+            import docplex.cp as docp
+            return True
+        except ModuleNotFoundError:
+            return False
+
+    @staticmethod
+    def executable_installed():
+        if shutil.which('cpoptimizer') is None:
+            return False
+        else:
+            return True
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: Model(), a CPMpy Model() (optional)
+        - subsolver: str, name of a subsolver (optional)
+        """
+        if not self.installed():
+            raise Exception("CPM_cpo: Install the python package 'docplex'")
+
+        if not self.executable_installed():
+            raise Exception("You need to install the CPLEX Optimization Studio to use this solver.")
+
+        import docplex.cp.model as dom
+        assert subsolver is None
+
+        # initialise the native solver object
+        # [GUIDELINE] we commonly use 3-letter abbrivations to refer to native objects:
+        #           OR-tools uses ort_solver, Gurobi grb_solver, Exact xct_solver...
+        self.cpo_model = dom.CpoModel()
+
+        # initialise everything else and post the constraints/objective
+        # [GUIDELINE] this superclass call should happen AFTER all solver-native objects are created.
+        #           internally, the constructor relies on __add__ which uses the above solver native object(s)
+        super().__init__(name="cpo", cpm_model=cpm_model)
+
+
+    def solve(self, time_limit=None, **kwargs):
+        """
+            Call the CP Optimizer solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            # [GUIDELINE] Please document key solver arguments that the user might wish to change
```

Comment:
some key solver args would be nice... e.g. log_output is one, maybe smth wrt nr of threads?